File: lib/src/main/java/com/sd/lib/vtrack/tracker/FViewTracker.java
package com.sd.lib.vtrack.tracker;

import android.view.View;
import android.view.ViewParent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;

/**
 * view的位置追踪
 */
public class FViewTracker implements ViewTracker {
    private WeakReference<View> mSource;
    private WeakReference<View> mTarget;

    private Location mSourceLocation;
    private Location mTargetLocation;

    private final int[] mLocationSourceParent = {0, 0};
    private final int[] mLocationTarget = {0, 0};

    private int mX;
    private int mY;
    private Position mPosition = Position.TopRight;

    private Callback mCallback;

    @Override
    public void setCallback(@Nullable Callback callback) {
        mCallback = callback;
    }

    @Override
    public void setSource(@Nullable View source) {
        final View old = getSource();
        if (old != source) {
            mSource = source == null ? null : new WeakReference<>(source);
            if (mCallback != null) {
                mCallback.onSourceChanged(old, source);
            }
        }
    }

    @Override
    public void setSourceLocation(@Nullable Location location) {
        mSourceLocation = location;
    }

    @Override
    public void setTarget(@Nullable View target) {
        final View old = getTarget();
        if (old != target) {
            mTarget = target == null ? null : new WeakReference<>(target);
            if (mCallback != null) {
                mCallback.onTargetChanged(old, target);
            }
        }
    }

    @Override
    public void setTargetLocation(@Nullable Location location) {
        mTargetLocation = location;
    }

    @Override
    public void setPosition(@NonNull Position position) {
        mPosition = position;
    }

    @Nullable
    @Override
    public View getSource() {
        return mSource == null ? null : mSource.get();
    }

    @Nullable
    @Override
    public View getTarget() {
        return mTarget == null ? null : mTarget.get();
    }

    @Override
    public final boolean update() {
        final Callback callback = mCallback;
        if (callback == null) {
            return false;
        }

        final View source = getSource();
        final View target = getTarget();
        if (source == null || target == null) {
            return false;
        }

        if (!callback.canUpdate(source, target)) {
            return false;
        }

        final ViewParent parent = source.getParent();
        if (!(parent instanceof View)) {
            return false;
        }

        if (!source.isAttachedToWindow() || !target.isAttachedToWindow()) {
            return false;
        }

        if (source.getWidth() <= 0 || source.getHeight() <= 0) {
            return false;
        }

        ((View) parent).getLocationOnScreen(mLocationSourceParent);
        target.getLocationOnScreen(mLocationTarget);

        switch (mPosition) {
            case TopLeft:
                layoutTopLeft(source, target);
                break;
            case TopCenter:
                layoutTopCenter(source, target);
                break;
            case TopRight:
                layoutTopRight(source, target);
                break;

            case LeftCenter:
                layoutLeftCenter(source, target);
                break;
            case Center:
                layoutCenter(source, target);
                break;
            case RightCenter:
                layoutRightCenter(source, target);
                break;

            case BottomLeft:
                layoutBottomLeft(source, target);
                break;
            case BottomCenter:
                layoutBottomCenter(source, target);
                break;
            case BottomRight:
                layoutBottomRight(source, target);
                break;

            case Left:
                layoutLeft(source, target);
                break;
            case Top:
                layoutTop(source, target);
                break;
            case Right:
                layoutRight(source, target);
                break;
            case Bottom:
                layoutBottom(source, target);
                break;
        }

        callback.onUpdate(mX, mY, source, target);
        return true;
    }

    private int getX_alignLeft() {
        return mLocationTarget[0] - mLocationSourceParent[0];
    }

    private int getX_alignRight(View source, View target) {
        return getX_alignLeft() + (target.getWidth() - source.getWidth());
    }

    private int getX_alignCenter(View source, View target) {
        return getX_alignLeft() + (target.getWidth() - source.getWidth()) / 2;
    }

    private int getY_alignTop() {
        return mLocationTarget[1] - mLocationSourceParent[1];
    }

    private int getY_alignBottom(View source, View target) {
        return getY_alignTop() + (target.getHeight() - source.getHeight());
    }

    private int getY_alignCenter(View source, View target) {
        return getY_alignTop() + (target.getHeight() - source.getHeight()) / 2;
    }

    //---------- position start----------

    private void layoutTopLeft(View source, View target) {
        mX = getX_alignLeft();
        mY = getY_alignTop();
    }

    private void layoutTopCenter(View source, View target) {
        mX = getX_alignCenter(source, target);
        mY = getY_alignTop();
    }

    private void layoutTopRight(View source, View target) {
        mX = getX_alignRight(source, target);
        mY = getY_alignTop();
    }


    private void layoutLeftCenter(View source, View target) {
        mX = getX_alignLeft();
        mY = getY_alignCenter(source, target);
    }

    private void layoutCenter(View source, View target) {
        mX = getX_alignCenter(source, target);
        mY = getY_alignCenter(source, target);
    }

    private void layoutRightCenter(View source, View target) {
        mX = getX_alignRight(source, target);
        mY = getY_alignCenter(source, target);
    }


    private void layoutBottomLeft(View source, View target) {
        mX = getX_alignLeft();
        mY = getY_alignBottom(source, target);
    }

    private void layoutBottomCenter(View source, View target) {
        mX = getX_alignCenter(source, target);
        mY = getY_alignBottom(source, target);
    }

    private void layoutBottomRight(View source, View target) {
        mX = getX_alignRight(source, target);
        mY = getY_alignBottom(source, target);
    }


    private void layoutLeft(View source, View target) {
        mX = getX_alignLeft();
        mY = source.getTop();
    }

    private void layoutTop(View source, View target) {
        mX = source.getLeft();
        mY = getY_alignTop();
    }

    private void layoutRight(View source, View target) {
        mX = getX_alignRight(source, target);
        mY = source.getTop();
    }

    private void layoutBottom(View source, View target) {
        mX = source.getLeft();
        mY = getY_alignBottom(source, target);
    }

    //---------- position end----------
}

File: lib/src/main/java/com/sd/lib/vtrack/tracker/FViewTracker.kt
package com.sd.lib.vtrack.tracker

import android.view.View
import com.sd.lib.vtrack.tracker.ViewTracker.*
import com.sd.lib.vtrack.tracker.location.WeakSourceViewLocationInfo
import com.sd.lib.vtrack.tracker.location.WeakViewLocationInfo

/**
 * 位置追踪
 */
class FViewTracker : ViewTracker {
    private val _locationSourceParent = intArrayOf(0, 0)
    private val _locationTarget = intArrayOf(0, 0)

    private var _x: Int? = 0
    private var _y: Int? = 0

    override var position = Position.TopRight
    override var sourceLocationInfo: SourceLocationInfo? = null
    override var targetLocationInfo: LocationInfo? = null
    override var callback: Callback? = null

    override var source: View?
        get() {
            val info = sourceLocationInfo ?: return null
            return if (info is ViewLocationInfo) info.view else null
        }
        set(value) {
            val info = sourceLocationInfo
            if (info is ViewLocationInfo) {
                info.view = value
            } else {
                sourceLocationInfo = object : WeakSourceViewLocationInfo() {
                    override fun onViewChanged(old: View?, view: View?) {
                        super.onViewChanged(old, view)
                        callback?.onSourceChanged(old, view)
                    }
                }.also { it.view = value }
            }
        }

    override var target: View?
        get() {
            val info = targetLocationInfo ?: return null
            return if (info is ViewLocationInfo) info.view else null
        }
        set(value) {
            val info = targetLocationInfo
            if (info is ViewLocationInfo) {
                info.view = value
            } else {
                targetLocationInfo = object : WeakViewLocationInfo() {
                    override fun onViewChanged(old: View?, view: View?) {
                        super.onViewChanged(old, view)
                        callback?.onTargetChanged(old, view)
                    }
                }.also { it.view = value }
            }
        }

    override fun update(): Boolean {
        // check null
        val callback = callback ?: return false
        val source = sourceLocationInfo ?: return false
        val target = targetLocationInfo ?: return false

        // check canUpdate
        if (!callback.canUpdate(source, target)) {
            return false
        }

        // check isReady
        if (!source.isReady) return false
        if (!target.isReady) return false

        // check parent
        val sourceParent = source.parentLocationInfo ?: return false
        if (!sourceParent.isReady) return false

        sourceParent.getCoordinate(_locationSourceParent)
        target.getCoordinate(_locationTarget)

        when (position) {
            Position.TopLeft -> layoutTopLeft(source, target)
            Position.TopCenter -> layoutTopCenter(source, target)
            Position.TopRight -> layoutTopRight(source, target)

            Position.LeftCenter -> layoutLeftCenter(source, target)
            Position.Center -> layoutCenter(source, target)
            Position.RightCenter -> layoutRightCenter(source, target)

            Position.BottomLeft -> layoutBottomLeft(source, target)
            Position.BottomCenter -> layoutBottomCenter(source, target)
            Position.BottomRight -> layoutBottomRight(source, target)

            Position.Left -> layoutLeft(source, target)
            Position.Top -> layoutTop(source, target)
            Position.Right -> layoutRight(source, target)
            Position.Bottom -> layoutBottom(source, target)
        }

        callback.onUpdate(_x, _y, source, target)
        return true
    }

    private fun getXAlignLeft(): Int {
        return _locationTarget[0] - _locationSourceParent[0]
    }

    private fun getXAlignRight(source: LocationInfo, target: LocationInfo): Int {
        return getXAlignLeft() + (target.width - source.width)
    }

    private fun getXAlignCenter(source: LocationInfo, target: LocationInfo): Int {
        return getXAlignLeft() + (target.width - source.width) / 2
    }

    private fun getYAlignTop(): Int {
        return _locationTarget[1] - _locationSourceParent[1]
    }

    private fun getYAlignBottom(source: LocationInfo, target: LocationInfo): Int {
        return getYAlignTop() + (target.height - source.height)
    }

    private fun getYAlignCenter(source: LocationInfo, target: LocationInfo): Int {
        return getYAlignTop() + (target.height - source.height) / 2
    }

    //---------- position start----------

    private fun layoutTopLeft(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignLeft()
        _y = getYAlignTop()
    }

    private fun layoutTopCenter(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignCenter(source, target)
        _y = getYAlignTop()
    }

    private fun layoutTopRight(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignRight(source, target)
        _y = getYAlignTop()
    }

    private fun layoutLeftCenter(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignLeft()
        _y = getYAlignCenter(source, target)
    }

    private fun layoutCenter(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignCenter(source, target)
        _y = getYAlignCenter(source, target)
    }

    private fun layoutRightCenter(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignRight(source, target)
        _y = getYAlignCenter(source, target)
    }

    private fun layoutBottomLeft(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignLeft()
        _y = getYAlignBottom(source, target)
    }

    private fun layoutBottomCenter(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignCenter(source, target)
        _y = getYAlignBottom(source, target)
    }

    private fun layoutBottomRight(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignRight(source, target)
        _y = getYAlignBottom(source, target)
    }

    private fun layoutLeft(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignLeft()
        _y = null
    }

    private fun layoutTop(source: LocationInfo, target: LocationInfo) {
        _x = null
        _y = getYAlignTop()
    }

    private fun layoutRight(source: LocationInfo, target: LocationInfo) {
        _x = getXAlignRight(source, target)
        _y = null
    }

    private fun layoutBottom(source: LocationInfo, target: LocationInfo) {
        _x = null
        _y = getYAlignBottom(source, target)
    }
}